### PR TITLE
Large internal rewrite, refactor - implemented wheel-based envs (v2.0 beta)

### PR DESCRIPTION
**Pull request by kylegibson**
* Terrarium no longer interfaces with pip and virtualenv **as libraries** - instead, all interfacing occurs at the command level
* Terrarium no longer creates a virtualenv bootstrap script (which was hella janky, tbh)
* Previously: A terrarium environment archive consisted of a nearly complete virtualenv, which had to be made relocatable
* Now: When building a new environment archive, terrarium will use or build wheels for every requirement. The archive is just the complete set of wheel files, plus the requirements.txt file
* Now: When installing a terrarium environment archive, a brand new virtualenv is created every time. Then the wheels are installed using the contained requirements.txt. Downloads from the pypi index and cache dir are disabled.
* Improved logging, including changing the meaning of `--quiet`, which now will completely silence all logging
* Improved error handling
* Improved parsing of requirements.txt files, **include handling -r or --requirement** options for including other requirements files: terrarium will create a flattened requirement structure
* Fixed S3 retry logic so it **actually retries** - previously, it would stop after the first failure.

### CODE REVIEW

File: tests/tests.py [hunk 1]
```diff
@@ -242,7 +250,7 @@ def test_install_empty_requirements(self):
 
         # Check for terrarium bootstrap script
```
Comment (bocribbz) [inline tests/tests.py]: this comment should probably be udpated or removed

File: tests/tests.py [hunk 2]
```diff
@@ -610,15 +619,15 @@ def test_restore_previously_backed_up_environment(self):
...
         )
+        self.assertEqual('', stderr)
         self.assertEqual(
-            output[1],
-            'Download archive failed\n'
-            'Failed to download bundle and download is required. '
-            'Refusing to build a new bundle.\n',
+            stdout.strip(),
+            '[ERROR] Failed to download environment and download is required. '
+            'Refusing to build a new environment.',
```
Comment (bocribbz) [inline tests/tests.py]: if this is an error, shoudnt it go into `stderr`?

File: terrarium/terrarium.py [hunk 3]
```diff
@@ -26,133 +24,55 @@
...
-        for arg in self.args.reqs:
-            if os.path.exists(arg):
-                with open(arg, 'r') as f:
-                    for line in f.readlines():
-                        line = line.strip()
-                        if line and not line.startswith('#'):
-                            lines.append(line)
+        for path in self.args.reqs:
+            if os.path.exists(path):
+                lines.extend(parse_requirements(path=path))
```
Comment (bocribbz) [inline terrarium/terrarium.py]: TIL

Comment (kylegibson) [reply terrarium/terrarium.py]: What did you learn?

File: tests/tests.py (hunk 2 above)
Comment (kylegibson) [reply tests/tests.py]: Ideally, yes... however python's logging module doesn't make it easy to send WARN/ERR to stderr and INFO/DEBUG to stdout. So, everything currently goes to stdout.

File: terrarium/terrarium.py (hunk 3 above)
Comment (bocribbz) [reply terrarium/terrarium.py]: `list.extend(generator)`
this would "generate" the generator and dump it in a list.

**Pull request merged**